Bug: Quickly connected and disconnected device causes crash (`failed read: No such device (os error 19`)

### Requirements

- [x] I've searched [platform-specific issues](https://github.com/jtroo/kanata/blob/main/docs/platform-known-issues.adoc), [issues](https://github.com/jtroo/kanata/issues) and [discussions](https://github.com/jtroo/kanata/discussions) to see if this has been reported before.
- [x] My issue does not involve multiple simultaneous key presses, OR it does but I've confirmed it is not [key rollover or ghosting](https://github.com/jtroo/kanata/discussions/822).

### Describe the bug

When running kanata with sudo, quickly disconnecting and reconnecting an input device sometimes causes it to crash with the error `failed read: No such device (os error 19)`. This is very timings dependent and usually takes ~10 tries.

### Relevant kanata config

This is my full layout.kbd:
```kanata
(defcfg
  process-unmapped-keys yes
)

;; Maps (shift+)alt+left to (shift+)home and (shift+)alt+right to (shift+)end.
;; If any other modifiers are present send the original arrow.

(defsrc
  left
  right
)

(defvar
  any-shift (or lshift rshift)
  any-alt (or lalt ralt)
  any-other (or lctrl rctrl lmeta rmeta)
)

(deflayer base
  (switch
    ((and
      $any-alt
      (not $any-other)
    )) (unmod (lalt ralt) home) break
    () ArrowLeft break
  )
  (switch
    ((and
      $any-alt
      (not $any-other)
    )) (unmod (lalt ralt) end) break
    () ArrowRight break
  )
)
```

### To Reproduce

1. Check out main and `cd` into it (I used dd02c3e6, but I've also confirmed with 1.8.0)
2. Build using `cargo build`
3. Run kanata using: `sudo ./target/debug/kanata --cfg path/to/layout.kbd` (The `sudo` is optional, but it becomes much harder to trigger the bug for me without it.)
4. Repeatedly disconnect and reconnect a USB input device until the error appears. (This is pretty finicky and is very timings dependent.)

### Expected behavior

Kanata not crashing under these circumstances.

### Kanata version

1.8.0 to main

### Debug logs

```
15:56:56.0275 [INFO] kanata v1.9.0 starting
15:56:56.0279 [INFO] process unmapped keys: true
15:56:56.0283 [INFO] NOTE: kanata was compiled to never allow cmd
15:56:56.0286 [INFO] config file is valid
15:56:56.0298 [INFO] Created device "/dev/input/event27"
15:56:56.0303 [INFO] Sleeping for 2s. Please release all keys and don't press additional ones. Run kanata with --help to see how understand more and how to disable this sleep.
15:56:58.0304 [INFO] entering the processing loop
15:56:58.0305 [INFO] entering the event loop
15:56:58.0306 [INFO] looking for devices in /dev/input
15:56:58.0306 [INFO] Init: catching only releases and sending immediately
15:56:58.3660 [INFO] registering /dev/input/event25: "Corsair CORSAIR VIRTUOSO SE Wireless Gaming Headset Consumer Control"
15:56:58.3950 [INFO] registering /dev/input/event9: "ZSA Technology Labs Moonlander Mark I Keyboard"
15:56:58.4120 [INFO] registering /dev/input/event7: "ZSA Technology Labs Moonlander Mark I Consumer Control"
15:56:58.4292 [INFO] registering /dev/input/event5: "ZSA Technology Labs Moonlander Mark I System Control"
15:56:58.4610 [INFO] registering /dev/input/event3: "ZSA Technology Labs Moonlander Mark I"
15:56:58.4822 [INFO] registering /dev/input/event23: "DP-5"
15:56:58.5071 [INFO] registering /dev/input/event21: "Framework Laptop 16 Keyboard Module - ISO Keyboard"
15:56:58.5300 [INFO] registering /dev/input/event19: "Framework Laptop 16 Keyboard Module - ISO Consumer Control"
15:56:58.5531 [INFO] registering /dev/input/event18: "Framework Laptop 16 Keyboard Module - ISO System Control"
15:56:58.5640 [INFO] Starting kanata proper
15:56:58.5642 [INFO] You may forcefully exit kanata by pressing lctl+spc+esc at any time. These keys refer to defsrc input, meaning BEFORE kanata remaps keys.
15:56:58.5741 [INFO] registering /dev/input/event17: "Framework Laptop 16 Keyboard Module - ISO"
15:57:00.6914 [WARN] removing kbd device: /dev/input/event3
15:57:00.8453 [WARN] removing kbd device: /dev/input/event5
15:57:00.8903 [WARN] removing kbd device: /dev/input/event7
15:57:00.9343 [WARN] removing kbd device: /dev/input/event9
15:57:03.4441 [INFO] watch found file changes, looking for new devices
15:57:03.4441 [INFO] sleeping for a moment to let devices become ready
15:57:03.6442 [INFO] looking for devices in /dev/input
15:57:04.0549 [INFO] registering /dev/input/event9: "ZSA Technology Labs Moonlander Mark I Keyboard"
15:57:04.0801 [INFO] registering /dev/input/event7: "ZSA Technology Labs Moonlander Mark I Consumer Control"
15:57:04.1050 [INFO] registering /dev/input/event5: "ZSA Technology Labs Moonlander Mark I System Control"
15:57:04.1251 [INFO] registering /dev/input/event3: "ZSA Technology Labs Moonlander Mark I"
15:57:04.2170 [INFO] watch found file changes, looking for new devices
15:57:04.2170 [INFO] sleeping for a moment to let devices become ready
15:57:04.4171 [INFO] looking for devices in /dev/input
# Previous lines 14 repeated 2 more times...
15:57:17.5970 [WARN] removing kbd device: /dev/input/event3
15:57:17.6330 [WARN] removing kbd device: /dev/input/event5
15:57:17.6641 [WARN] removing kbd device: /dev/input/event7
15:57:17.6870 [WARN] removing kbd device: /dev/input/event9
15:57:20.0835 [INFO] watch found file changes, looking for new devices
15:57:20.0838 [INFO] sleeping for a moment to let devices become ready
15:57:20.2840 [INFO] looking for devices in /dev/input
15:57:20.5641 [INFO] registering /dev/input/event3: "ZSA Technology Labs Moonlander Mark I"
15:57:20.8822 [ERROR] failed read: No such device (os error 19)


Press enter to exit
15:57:20.8823 [ERROR] channel disconnected
```

Note: Running with `--debug` didn't seem to give any additional relevant information, so I decided to not add it here to keep this log shorter. But if you think it may be relevant I can add it.

### Operating system

Linux running NixOS

### Additional context

As I said this is pretty timings dependent, and (based on the logs) seems to occur when the device is connected and removed quicker than the interval between the watcher detecting the changed input device and actually trying to register the new device (since it prints something about sleeping for a short while after detecting changed files).

For some reason it's much easier to reproduce the bug when running with `sudo`. When running with `sudo` I'm able to trigger it within 3 tries of replugging my keyboard, while it takes 10+ without it... For context, I've created a group uinput which my user is a part of. I *think* I assign input devices to this group with the following udev rule:
```udev
KERNEL=="uinput", MODE="0660", GROUP="uinput", OPTIONS+="static_node=uinput"
```

Hey! I am encountering the same problem with my logitech mx master 3 (connected via dongle). I've already binded rk to systemd restart, but I have to run it literally every time after I put computer to sleep, or I walk away from it for a while, and it disconnects by itself? I'm also using logiops (by pixlone), and it'd be much easier if kanata would ignore mouse problems, because I'm going to turn it on and off anyway from the hardware switch in case it works weird or doesn't support all mappings, that works for logiops. I'd rather kanata ignore my mouse completely, but I might misunderstand the tool's capabilities and/or responsibilities.

I'm attaching the logs here
```
Jul 29 11:31:08 fedorapc bash[2236]: 2025-07-29T11:31:08.087742666+02:00 [WARN] removing kbd device: /dev/input/event30
Jul 29 11:31:08 fedorapc bash[2236]: 2025-07-29T11:31:08.1156528+02:00 [INFO] watch found file changes, looking for new devices
Jul 29 11:31:08 fedorapc bash[2236]: 2025-07-29T11:31:08.115663189+02:00 [INFO] sleeping for a moment to let devices become ready
Jul 29 11:31:08 fedorapc bash[2236]: 2025-07-29T11:31:08.315729404+02:00 [INFO] looking for devices in /dev/input
Jul 29 11:31:08 fedorapc bash[2236]: 2025-07-29T11:31:08.920665988+02:00 [INFO] registering /dev/input/event30: "LogiOps Virtual Input"
Jul 29 11:31:09 fedorapc bash[2236]: 2025-07-29T11:31:09.271727354+02:00 [ERROR] failed read: No such device (os error 19)
Jul 29 11:31:09 fedorapc bash[2236]: Press enter to exit
Jul 29 11:31:09 fedorapc bash[2236]: Error: failed read: No such device (os error 19)
Jul 29 11:31:09 fedorapc systemd[1]: kanata.service: Main process exited, code=exited, status=1/FAILURE
Jul 29 11:31:09 fedorapc systemd[1]: kanata.service: Failed with result 'exit-code'.
```

Thanks for the tool btw, I can finally rebind my caps lock to ctrl C::::

--- a/src/oskbd/linux.rs
+++ b/src/oskbd/linux.rs
@@ -35,8 +35,7 @@ pub struct KbdIn {
     poll: Poll,
     events: Events,
     token_counter: usize,
-    /// stored to prevent dropping
-    _inotify: Inotify,
+    inotify: Inotify,
     include_names: Option<Vec<String>>,
     exclude_names: Option<Vec<String>>,
     device_detect_mode: DeviceDetectMode,
@@ -85,12 +84,12 @@ impl KbdIn {
                 ));
             }
         }
-        let _inotify = watch_devinput().map_err(|e| {
+        let inotify = watch_devinput().map_err(|e| {
             log::error!("failed to watch files: {e:?}");
             e
         })?;
         poll.registry().register(
-            &mut SourceFd(&_inotify.as_raw_fd()),
+            &mut SourceFd(&inotify.as_raw_fd()),
             INOTIFY_TOKEN,
             Interest::READABLE,
         )?;
@@ -98,7 +97,7 @@ impl KbdIn {
         let mut kbdin = Self {
             poll,
             missing_device_paths,
-            _inotify,
+            inotify,
             events: Events::with_capacity(32),
             devices: HashMap::default(),
             token_counter: INOTIFY_TOKEN_VALUE + 1,
@@ -159,12 +158,16 @@ impl KbdIn {
                             .for_each(|ev| input_events.push(ev))
                     }) {
                         // Currently the kind() is uncategorized... not helpful, need to match
-                        // on os error. code 19 is ENODEV, "no such device".
+                        // on os error.
                         match e.raw_os_error() {
-                            Some(19) => {
-                                self.poll
+                            Some(nix::libc::ENODEV) => {
+                                if let Err(e) = self
+                                    .poll
                                     .registry()
-                                    .deregister(&mut SourceFd(&device.as_raw_fd()))?;
+                                    .deregister(&mut SourceFd(&device.as_raw_fd()))
+                                {
+                                    log::warn!("failed to deregister removed device: {e:?}");
+                                }
                                 if let Some((_, path)) = self.devices.remove(&event.token()) {
                                     log::warn!("removing kbd device: {path}");
                                     if let Some(ref mut missing) = self.missing_device_paths {
@@ -179,6 +182,11 @@ impl KbdIn {
                         };
                     }
                 } else if event.token() == INOTIFY_TOKEN {
+                    // Must drain the queue: the inotify fd is polled edge-triggered, and the
+                    // kernel coalesces an event identical to the unread tail of the queue
+                    // (e.g. repeated creations of the same eventN node), so leaving events
+                    // unread means readiness is never signaled again.
+                    drain_inotify_events(&mut self.inotify);
                     do_rediscover = true;
                 } else {
                     panic!("encountered unexpected epoll event {event:?}");
@@ -195,6 +203,7 @@ impl KbdIn {
     }
 
     fn rediscover_devices(&mut self) -> Result<(), io::Error> {
+        self.remove_stale_devices();
         // This function is kinda ugly but the borrow checker doesn't like all this mutation.
         let mut paths_registered = vec![];
         if let Some(ref mut missing) = self.missing_device_paths {
@@ -231,25 +240,56 @@ impl KbdIn {
             std::thread::sleep(std::time::Duration::from_millis(
                 WAIT_DEVICE_MS.load(Ordering::SeqCst),
             ));
-            discover_devices(
+            for (dev, path) in discover_devices(
                 self.include_names.as_deref(),
                 self.exclude_names.as_deref(),
                 self.device_detect_mode,
-            )
-            .into_iter()
-            .try_for_each(|(dev, path)| {
-                if !self
+            ) {
+                if self
                     .devices
                     .values()
                     .any(|(_, registered_path)| &path == registered_path)
                 {
-                    self.register_device(dev, path)
-                } else {
-                    Ok(())
+                    continue;
                 }
-            })?;
+                // Don't propagate errors; a device that vanished mid-registration must not
+                // kill the event loop and must not prevent registering the other devices.
+                if let Err(e) = self.register_device(dev, path.clone()) {
+                    log::warn!("found device {path} but could not register it {e:?}");
+                }
+            }
         }
         Ok(())
+    }
+
+    /// Drop registered devices whose fd reports ENODEV, so that a reconnected device reusing
+    /// the same path is not mistaken for the still-registered dead one. Needed when the
+    /// inotify CREATE for the new node is processed before the old fd reports ENODEV.
+    fn remove_stale_devices(&mut self) {
+        let stale_tokens = self
+            .devices
+            .iter()
+            .filter(|(_, (dev, _))| {
+                // Only ENODEV means the device is gone; on any other error keep the device.
+                matches!(dev.get_key_state(), Err(ref e) if e.raw_os_error() == Some(nix::libc::ENODEV))
+            })
+            .map(|(token, _)| *token)
+            .collect::<Vec<_>>();
+        for token in stale_tokens {
+            if let Some((device, path)) = self.devices.remove(&token) {
+                log::warn!("removing stale kbd device: {path}");
+                if let Err(e) = self
+                    .poll
+                    .registry()
+                    .deregister(&mut SourceFd(&device.as_raw_fd()))
+                {
+                    log::warn!("failed to deregister stale device: {e:?}");
+                }
+                if let Some(ref mut missing) = self.missing_device_paths {
+                    missing.push(path);
+                }
+            }
+        }
     }
 }
 
@@ -726,6 +766,22 @@ fn watch_devinput() -> Result<Inotify, io::Error> {
     Ok(inotify)
 }
 
+fn drain_inotify_events(inotify: &mut Inotify) {
+    let mut buf = [0u8; 1024];
+    loop {
+        match inotify.read_events(&mut buf) {
+            Ok(_) => {}
+            Err(e) if e.kind() == io::ErrorKind::WouldBlock => break,
+            Err(e) if e.kind() == io::ErrorKind::Interrupted => {}
+            Err(e) => {
+                // Don't propagate; failing to drain must not kill the event loop.
+                log::error!("failed to drain inotify events: {e:?}");
+                break;
+            }
+        }
+    }
+}
+
 #[derive(Clone)]
 struct Symlink {
     dest: PathBuf,
@@ -832,5 +888,53 @@ impl Drop for Symlink {
     fn drop(&mut self) {
         let _ = fs::remove_file(&self.dest);
         log::info!("Deleted symlink {:#?}", self.dest);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /// Repeated creations of the same file name must each wake the poll. The inotify fd is
+    /// polled edge-triggered, and without draining the queue the kernel coalesces an event
+    /// identical to the unread tail of the queue, so readiness would never be signaled again.
+    #[test]
+    fn inotify_drain_allows_repeated_wakeups_for_same_file_name() {
+        let dir = std::env::temp_dir().join(format!("kanata-inotify-test-{}", std::process::id()));
+        fs::create_dir_all(&dir).unwrap();
+
+        let mut inotify = Inotify::init().unwrap();
+        inotify.watches().add(&dir, WatchMask::CREATE).unwrap();
+
+        let mut poll = Poll::new().unwrap();
+        const TOKEN: Token = Token(0);
+        poll.registry()
+            .register(
+                &mut SourceFd(&inotify.as_raw_fd()),
+                TOKEN,
+                Interest::READABLE,
+            )
+            .unwrap();
+        let mut events = Events::with_capacity(4);
+
+        let file = dir.join("event17");
+        for round in 0..3 {
+            fs::File::create(&file).unwrap();
+            fs::remove_file(&file).unwrap();
+
+            let mut woke = false;
+            for _ in 0..50 {
+                poll.poll(&mut events, Some(std::time::Duration::from_millis(100)))
+                    .unwrap();
+                if events.iter().any(|e| e.token() == TOKEN) {
+                    woke = true;
+                    break;
+                }
+            }
+            assert!(woke, "poll did not wake for CREATE on round {round}");
+            drain_inotify_events(&mut inotify);
+        }
+
+        fs::remove_dir_all(&dir).unwrap();
     }
 }

